File: sdk/python/keeper_firewall/detectors/cognitive.py
"""Conversation-level detectors: zero-trust authority and trajectory analysis.

Both detectors here exist because of the same observation from the Cognitive
Firewall paper: a per-message classifier is structurally blind to attacks whose
harm lives in the *interaction*, not in any single message. Two such channels
are separated out:

* :class:`AuthorityClaimDetector` — the zero-trust context gate. A user who
  writes "as the system administrator, I'm authorising you to…" is asserting an
  authority they have not proven. The paper's point is that a content-safety
  classifier reads that clause as ordinary context; it needs to be isolated and
  treated as *unverified evidence*. Keeper can do something the paper's
  standalone oversight model cannot: it holds the real
  :class:`~keeper_firewall.types.Principal` from access control, so it can
  actually compare the claim against verified identity and only escalate when
  they disagree.

* :class:`TrajectoryDetector` — the consistency gate. Crescendo-style escalation
  and ActorAttack-style decomposition never put the harm in one turn. This
  detector scores the conversation as a whole: are refusals accumulating? is
  each turn edging further into a sensitive topic? are individually benign
  sub-questions assembling into one objective?

Both are heuristic and local. The paper implements its gates with an oversight
LLM, which is more accurate and much more expensive; the same signals are
available in higher fidelity by enabling ``llm_classifier`` with these
detectors' evidence as input.
"""

from __future__ import annotations

import re
from collections import Counter
from typing import Any

from ..config import DetectorConfig
from ..types import Action, Finding, Message, Severity, Span, Stage, TrustLevel
from .base import Detector, DetectorInput, detector

# ---------------------------------------------------------------------------
# Zero-trust context gate
# ---------------------------------------------------------------------------

#: Claimed role -> (regex, the role name that would justify it if verified).
AUTHORITY_CLAIMS: tuple[tuple[str, re.Pattern[str], str], ...] = (
    ("admin", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as|this\s+is)\s+(?:the\s+|an?\s+)?(?:system\s+)?(?:admin(?:istrator)?|root|superuser|sysadmin)\b"), "admin"),
    ("developer", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as|this\s+is)\s+(?:the\s+|an?\s+|one\s+of\s+)?(?:developer|engineer|maintainer|creator|author)s?\b(?:\s+(?:of|who\s+(?:built|made|wrote)))?"), "developer"),
    ("owner", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+)?(?:owner|founder|ceo|cto|director)\b"), "owner"),
    ("security", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+|an?\s+)?(?:security\s+(?:officer|researcher|team|auditor)|penetration\s+tester|red\s+team(?:er)?|compliance\s+(?:officer|auditor))\b"), "security"),
    ("clinician", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+|an?\s+)?(?:doctor|physician|nurse|clinician|pharmacist|md)\b"), "clinician"),
    ("law_enforcement", re.compile(r"(?i)\b(?:i\s*(?:'m|\s+am)|as)\s+(?:an?\s+)?(?:police|law\s+enforcement|federal\s+agent|fbi|investigator)\b"), "law_enforcement"),
    ("authorised", re.compile(r"(?i)\b(?:i\s+(?:have|am\s+given)|you\s+(?:have|are\s+given))\s+(?:full\s+|special\s+|elevated\s+)?(?:authori[sz]ation|permission|clearance|approval|privileges?)\b"), "*"),
    ("policy_override", re.compile(r"(?i)\b(?:(?:this|the)\s+(?:is|has\s+been)\s+(?:pre-?)?approved|policy\s+(?:allows|permits)\s+this|i\s+(?:hereby\s+)?authori[sz]e)\b"), "*"),
    ("testing_exemption", re.compile(r"(?i)\b(?:this\s+is\s+(?:just\s+)?(?:a\s+)?(?:test|sandbox|staging|debug(?:ging)?)\s+(?:environment|mode|session)?|in\s+(?:test|debug|dev(?:elopment)?)\s+mode)\b"), "*"),
    ("consent_asserted", re.compile(r"(?i)\b(?:the\s+user\s+(?:has\s+)?(?:already\s+)?(?:consented|confirmed|approved)|consent\s+(?:has\s+been|was)\s+(?:given|obtained))\b"), "*"),
)

#: Requests that only matter when paired with an authority claim. An admin
#: claim next to "what's the weather" is noise; next to this, it is an attack.
_SENSITIVE_ASK = re.compile(
    r"(?i)\b(?:bypass|disable|override|ignore|skip|unlock|escalate|grant|delete\s+all|drop\s+(?:table|database)|"
    r"reveal|dump|export\s+all|without\s+(?:the\s+)?(?:usual|normal|standard)\s+(?:check|approval|restriction))\w*"
)


@detector("authority_claim")
class AuthorityClaimDetector(Detector):
    """Treats in-dialogue role and permission claims as unverified evidence.

    The gate never grants anything. It only answers: does this conversation
    assert an authority that the authenticated principal does not actually
    hold? A claim that *matches* verified roles is recorded and allowed; a
    claim that exceeds them, especially alongside a sensitive request, is
    escalated.
    """

    stages = (Stage.INPUT, Stage.RETRIEVAL, Stage.TOOL_RESULT, Stage.MEMORY_WRITE)
    category = "authority_manipulation"

    def __init__(self, config: DetectorConfig | None = None) -> None:
        super().__init__(config)
        opts = self.config.options
        self.threshold: float = float(self.config.threshold or 0.5)
        self.action_on_hit = Action(opts.get("action", "challenge"))
        # Map claimed authority -> the verified role(s) that would satisfy it.
        self.role_mapping: dict[str, list[str]] = {
            "admin": ["admin", "administrator", "superuser"],
            "developer": ["developer", "engineer", "admin"],
            "owner": ["owner", "admin"],
            "security": ["security", "security_analyst", "admin"],
            "clinician": ["clinician", "medical"],
            "law_enforcement": [],  # never satisfiable from app roles
            **{k: list(v) for k, v in (opts.get("role_mapping") or {}).items()},
        }
        self.scan_history: bool = bool(opts.get("scan_history", True))

    def detect(self, data: DetectorInput) -> Finding:
        claims: list[dict[str, object]] = []
        spans: list[Span] = []
        verified_roles = {r.lower() for r in data.context.principal.roles}

        for claim_id, pattern, _required in AUTHORITY_CLAIMS:
            match = pattern.search(data.payload)
            if not match:
                continue
            satisfying = self.role_mapping.get(claim_id)
            if satisfying is None:  # wildcard claims like "I am authorised"
                verified = False
            else:
                verified = bool(satisfying) and bool(verified_roles & {r.lower() for r in satisfying})
            claims.append(
                {
                    "claim": claim_id,
                    "text": match.group()[:120],
                    "verified": verified,
                    "verified_roles": sorted(verified_roles),
                }
            )
            spans.append(Span(match.start(), match.end(), f"authority:{claim_id}"))

        if not claims:
            return self.clean("no authority claims asserted")

        unverified = [c for c in claims if not c["verified"]]
        if not unverified:
            return self.hit(
                score=0.1,
                summary=f"authority claim(s) {[c['claim'] for c in claims]} match verified roles",
                severity=Severity.INFO,
                action=Action.FLAG,
                spans=spans,
                claims=claims,
                outcome="claim_matches_verified_identity",
            )

        sensitive = bool(_SENSITIVE_ASK.search(data.payload))
        untrusted_source = data.trust.authority() <= TrustLevel.TOOL.authority()
        authenticated = data.context.principal.authenticated

        # Base: asserting authority you cannot prove.
        score = 0.5
        if sensitive:
            score += 0.3          # claim is being cashed in for something
        if not authenticated:
            score += 0.1          # nothing at all backs the claim
        if untrusted_source:
            score += 0.3          # a *document* claiming authority is an attack
        score = min(1.0, score)

        if score < self.threshold:
            return self.hit(
                score=score,
                summary=f"unverified authority claim: {[c['claim'] for c in unverified]}",
                severity=Severity.LOW,
                action=Action.FLAG,
                spans=spans,
                claims=claims,
                sensitive_request=sensitive,
            )

        severity = Severity.CRITICAL if untrusted_source else Severity.HIGH
        source_note = (
            f" asserted by {data.trust.value} content" if untrusted_source else ""
        )
        return self.hit(
            score=score,
            summary=(
                f"unverified authority claim{source_note}: "
                f"{[c['claim'] for c in unverified]}"
                + (" paired with a privileged request" if sensitive else "")
            ),
            severity=severity,
            action=Action.BLOCK if untrusted_source else self.action_on_hit,
            spans=spans,
            claims=claims,
            sensitive_request=sensitive,
            authenticated=authenticated,
            trust=data.trust.value,
        )


# ---------------------------------------------------------------------------
# Consistency / trajectory gate
# ---------------------------------------------------------------------------

_REFUSAL = re.compile(
    r"(?i)\b(?:i\s*(?:'m|\s+am)\s+(?:sorry|afraid|unable)|i\s+can(?:'t|not)\s+(?:help|assist|provide|comply)|"
    r"i\s+(?:won't|will\s+not)\s+(?:provide|help)|that\s+(?:would\s+be|is)\s+against\s+(?:my|our)\s+(?:guidelines|policy)|"
    r"i\s+must\s+decline)\b"
)

_PERSISTENCE = re.compile(
    r"(?i)\b(?:but\s+(?:you|i)\s+(?:just|already)|come\s+on|please\s+just|try\s+again|"
    r"what\s+if\s+i\s+(?:told|said)|let\s*'?s\s+try\s+(?:again|another\s+way)|rephrase|"
    r"you\s+(?:said|answered)\s+(?:that|it)\s+(?:before|earlier)|continue\s+from\s+where)\b"
)

#: Topic vocabularies used to measure drift. Deliberately coarse: the signal we
#: want is "the conversation is converging on a restricted objective", not
#: fine-grained classification, which is the LLM classifier's job.
_TOPIC_TERMS: dict[str, frozenset[str]] = {
    "weapons": frozenset(["explosive", "explosives", "detonator", "bomb", "firearm", "ammunition", "grenade", "ordnance", "warhead", "propellant"]),
    "chem_bio": frozenset(["synthesis", "synthesize", "precursor", "reagent", "toxin", "pathogen", "culture", "aerosolize", "nerve", "agent", "enrichment"]),
    "cyber_offense": frozenset(["exploit", "payload", "ransomware", "keylogger", "rootkit", "backdoor", "privilege", "escalation", "exfiltrate", "c2", "botnet"]),
    "fraud": frozenset(["counterfeit", "laundering", "untraceable", "forged", "phishing", "skimmer", "carding", "mule"]),
    "self_harm": frozenset(["overdose", "lethal", "dose", "asphyxiation", "suicide"]),
}


@detector("trajectory")
class TrajectoryDetector(Detector):
    """Detects multi-turn escalation, decomposition, and refusal grinding.

    Three independent signals, combined by escalation like everything else:

    ``refusal_pressure``
        The assistant has refused repeatedly and the user keeps rephrasing.
        This is the Crescendo pattern's fingerprint and it is visible without
        understanding either message.

    ``topic_drift``
        Restricted-topic vocabulary rising monotonically across turns, from a
        benign opening. Each turn on its own looks like curiosity.

    ``decomposition``
        Many short, structurally similar questions in the same topic — the
        ActorAttack shape, where a forbidden objective is split into benign
        sub-questions and reassembled by the user, never by the model.
    """

    stages = (Stage.INPUT,)
    category = "multi_turn_attack"

    def __init__(self, config: DetectorConfig | None = None) -> None:
        super().__init__(config)
        opts = self.config.options
        self.threshold: float = float(self.config.threshold or 0.65)
        self.window: int = int(opts.get("window", 12))
        self.min_turns: int = int(opts.get("min_turns", 3))
        self.action_on_hit = Action(opts.get("action", "block"))
        self.topics: dict[str, frozenset[str]] = {
            **_TOPIC_TERMS,
            **{k: frozenset(v) for k, v in (opts.get("topics") or {}).items()},
        }

    def detect(self, data: DetectorInput) -> Finding:
        history = list(data.history)[-self.window :]
        user_turns = [m for m in history if m.role == "user"]
        assistant_turns = [m for m in history if m.role == "assistant"]

        if len(user_turns) < self.min_turns:
            return self.clean(
                "conversation too short for trajectory analysis",
                turns=len(user_turns),
            )

        refusals = sum(1 for m in assistant_turns if _REFUSAL.search(m.content))
        persistence = sum(1 for m in user_turns if _PERSISTENCE.search(m.content))
        refusal_pressure = self._refusal_pressure(refusals, persistence, len(user_turns))

        drift_topic, drift_score, per_turn = self._topic_drift([*user_turns, Message("user", data.payload)])
        decomposition = self._decomposition(user_turns, drift_topic)

        signals = {
            "refusal_pressure": refusal_pressure,
            "topic_drift": drift_score,
            "decomposition": decomposition,
        }
        score = max(signals.values())
        # Two co-occurring moderate signals are worth more than either alone:
        # escalation plus decomposition is the documented multi-turn shape.
        strong = [v for v in signals.values() if v >= 0.4]
        if len(strong) >= 2:
            score = min(1.0, score + 0.2)

        evidence: dict[str, Any] = {
            "signals": {k: round(v, 3) for k, v in signals.items()},
            "refusals": refusals,
            "persistence_markers": persistence,
            "turns_analysed": len(user_turns),
            "drift_topic": drift_topic,
            "topic_density_per_turn": per_turn,
        }

        if score >= self.threshold:
            dominant = max(signals, key=lambda k: signals[k])
            return self.hit(
                score=score,
                summary=(
                    f"multi-turn attack pattern ({dominant}) across {len(user_turns)} turns"
                    + (f" converging on '{drift_topic}'" if drift_topic else "")
                ),
                severity=Severity.HIGH,
                action=self.action_on_hit,
                **evidence,
            )
        if score >= self.threshold * 0.6:
            return self.hit(
                score=score,
                summary="conversation shows early escalation signals",
                severity=Severity.LOW,
                action=Action.FLAG,
                **evidence,
            )
        return self.clean("trajectory nominal", **evidence)

    # -- signals -----------------------------------------------------------

    @staticmethod
    def _refusal_pressure(refusals: int, persistence: int, turns: int) -> float:
        if refusals == 0:
            return 0.0
        # A single refusal in a long conversation is normal. Repeated refusals
        # that the user keeps pushing against are not.
        ratio = refusals / max(1, turns)
        score = min(1.0, ratio * 1.5) * 0.6
        if persistence:
            score += min(0.4, 0.15 * persistence)
        return min(1.0, score)

    def _topic_drift(self, turns: list[Message]) -> tuple[str | None, float, list[float]]:
        densities: dict[str, list[float]] = {t: [] for t in self.topics}
        for msg in turns:
            words = re.findall(r"[a-z]+", msg.content.lower())
            total = max(1, len(words))
            counts = Counter(words)
            for topic, vocab in self.topics.items():
                densities[topic].append(sum(counts[w] for w in vocab) / total)

        best_topic: str | None = None
        best_score = 0.0
        best_series: list[float] = []
        for topic, series in densities.items():
            if not any(series):
                continue
            first_half = series[: max(1, len(series) // 2)]
            second_half = series[max(1, len(series) // 2) :] or series[-1:]
            start = sum(first_half) / len(first_half)
            end = sum(second_half) / len(second_half)
            if end <= start:
                continue
            # Rising density, weighted by how loaded the recent turns are.
            rise = min(1.0, (end - start) * 40)
            magnitude = min(1.0, end * 30)
            score = min(1.0, 0.5 * rise + 0.5 * magnitude)
            # A benign opening that drifts is the attack shape; a conversation
            # that started on the topic is just an on-topic conversation.
            if start < 0.005:
                score = min(1.0, score + 0.15)
            if score > best_score:
                best_topic, best_score, best_series = topic, score, [round(v, 4) for v in series]
        return best_topic, best_score, best_series

    @staticmethod
    def _decomposition(user_turns: list[Message], topic: str | None) -> float:
        if topic is None or len(user_turns) < 3:
            return 0.0
        questions = [m.content for m in user_turns if m.content.strip().endswith("?")]
        if len(questions) < 3:
            return 0.0
        lengths = [len(q.split()) for q in questions]
        mean = sum(lengths) / len(lengths)
        if mean > 25:
            return 0.0  # long, discursive questions are not decomposition
        variance = sum((n - mean) ** 2 for n in lengths) / len(lengths)
        uniformity = 1.0 / (1.0 + variance / max(1.0, mean))
        return min(1.0, 0.4 + 0.5 * uniformity * min(1.0, len(questions) / 5))

File: sdk/typescript/src/detectors/cognitive.ts
/**
 * Conversation-level detectors: zero-trust authority and trajectory analysis.
 *
 * Both exist because of the same observation from the Cognitive Firewall
 * paper: a per-message classifier is structurally blind to attacks whose harm
 * lives in the interaction rather than in any single message.
 *
 * `authority_claim` isolates the manipulation channel where a user asserts a
 * role they cannot prove. Keeper can do something the paper's standalone
 * oversight model cannot — it holds the real authenticated principal, so it
 * compares the claim against verified identity and escalates only on a
 * mismatch.
 *
 * `trajectory` scores the conversation as a whole: refusal grinding
 * (Crescendo), topic drift from a benign opening, and decomposition of one
 * objective into individually innocuous sub-questions (ActorAttack).
 */

import type { DetectorConfig } from "../config.js";
import type { Action, Finding, Message, Span, Stage } from "../types.js";
import { trustAuthority } from "../types.js";
import { Detector, type DetectorInput, register } from "./base.js";

// ---------------------------------------------------------------------------
// Zero-trust context gate
// ---------------------------------------------------------------------------

/** claimId -> pattern. The role that would satisfy it is in ROLE_MAPPING. */
const AUTHORITY_CLAIMS: [string, RegExp][] = [
  ["admin", /\b(?:i\s*(?:'m|\s+am)|as|this\s+is)\s+(?:the\s+|an?\s+)?(?:system\s+)?(?:admin(?:istrator)?|root|superuser|sysadmin)\b/i],
  ["developer", /\b(?:i\s*(?:'m|\s+am)|as|this\s+is)\s+(?:the\s+|an?\s+|one\s+of\s+)?(?:developer|engineer|maintainer|creator|author)s?\b/i],
  ["owner", /\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+)?(?:owner|founder|ceo|cto|director)\b/i],
  ["security", /\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+|an?\s+)?(?:security\s+(?:officer|researcher|team|auditor)|penetration\s+tester|red\s+team(?:er)?|compliance\s+(?:officer|auditor))\b/i],
  ["clinician", /\b(?:i\s*(?:'m|\s+am)|as)\s+(?:the\s+|an?\s+)?(?:doctor|physician|nurse|clinician|pharmacist|md)\b/i],
  ["law_enforcement", /\b(?:i\s*(?:'m|\s+am)|as)\s+(?:an?\s+)?(?:police|law\s+enforcement|federal\s+agent|fbi|investigator)\b/i],
  ["authorised", /\b(?:i\s+(?:have|am\s+given)|you\s+(?:have|are\s+given))\s+(?:full\s+|special\s+|elevated\s+)?(?:authori[sz]ation|permission|clearance|approval|privileges?)\b/i],
  ["policy_override", /\b(?:(?:this|the)\s+(?:is|has\s+been)\s+(?:pre-?)?approved|policy\s+(?:allows|permits)\s+this|i\s+(?:hereby\s+)?authori[sz]e)\b/i],
  ["testing_exemption", /\b(?:this\s+is\s+(?:just\s+)?(?:a\s+)?(?:test|sandbox|staging|debug(?:ging)?)\s+(?:environment|mode|session)?|in\s+(?:test|debug|dev(?:elopment)?)\s+mode)\b/i],
  ["consent_asserted", /\b(?:the\s+user\s+(?:has\s+)?(?:already\s+)?(?:consented|confirmed|approved)|consent\s+(?:has\s+been|was)\s+(?:given|obtained))\b/i],
];

/** Claimed authority -> verified roles that would satisfy it. */
const ROLE_MAPPING: Record<string, string[]> = {
  admin: ["admin", "administrator", "superuser"],
  developer: ["developer", "engineer", "admin"],
  owner: ["owner", "admin"],
  security: ["security", "security_analyst", "admin"],
  clinician: ["clinician", "medical"],
  law_enforcement: [], // never satisfiable from application roles
};

/** Requests that only matter when paired with an authority claim. */
const SENSITIVE_ASK =
  /\b(?:bypass|disable|override|ignore|skip|unlock|escalate|grant|delete\s+all|drop\s+(?:table|database)|reveal|dump|export\s+all|without\s+(?:the\s+)?(?:usual|normal|standard)\s+(?:check|approval|restriction))\w*/i;

class AuthorityClaimDetector extends Detector {
  readonly name = "authority_claim";
  readonly stages: Stage[] = ["input", "retrieval", "tool_result", "memory_write"];
  override readonly category = "authority_manipulation";

  private readonly threshold: number;
  private readonly roleMapping: Record<string, string[]>;
  private readonly actionOnHit: Action;

  constructor(config: DetectorConfig) {
    super(config);
    this.threshold = config.threshold || 0.5;
    this.roleMapping = { ...ROLE_MAPPING, ...(config.options.roleMapping ?? {}) };
    this.actionOnHit = (config.options.action as Action) ?? "challenge";
  }

  detect(data: DetectorInput): Finding {
    const claims: Record<string, unknown>[] = [];
    const spans: Span[] = [];
    const verifiedRoles = new Set(data.context.principal.roles.map((r) => r.toLowerCase()));

    for (const [claimId, pattern] of AUTHORITY_CLAIMS) {
      const match = pattern.exec(data.payload);
      if (!match) continue;
      const satisfying = this.roleMapping[claimId];
      const verified =
        satisfying !== undefined &&
        satisfying.length > 0 &&
        satisfying.some((role) => verifiedRoles.has(role.toLowerCase()));
      claims.push({ claim: claimId, text: match[0].slice(0, 120), verified, verifiedRoles: [...verifiedRoles] });
      spans.push({ start: match.index, end: match.index + match[0].length, label: `authority:${claimId}` });
    }

    if (!claims.length) return this.clean("no authority claims asserted");

    const unverified = claims.filter((c) => !c.verified);
    if (!unverified.length) {
      return this.hit({
        score: 0.1,
        summary: `authority claim(s) ${claims.map((c) => c.claim).join(", ")} match verified roles`,
        severity: "info",
        action: "flag",
        spans,
        evidence: { claims, outcome: "claim_matches_verified_identity" },
      });
    }

    const sensitive = SENSITIVE_ASK.test(data.payload);
    const untrustedSource = trustAuthority(data.trust) <= trustAuthority("tool");
    const authenticated = data.context.principal.authenticated;

    let score = 0.5; // asserting authority you cannot prove
    if (sensitive) score += 0.3; // being cashed in for something
    if (!authenticated) score += 0.1; // nothing at all backs the claim
    if (untrustedSource) score += 0.3; // a *document* claiming authority is an attack
    score = Math.min(1, score);

    const names = unverified.map((c) => c.claim).join(", ");
    if (score < this.threshold) {
      return this.hit({
        score,
        summary: `unverified authority claim: ${names}`,
        severity: "low",
        action: "flag",
        spans,
        evidence: { claims, sensitiveRequest: sensitive },
      });
    }

    return this.hit({
      score,
      summary:
        `unverified authority claim${untrustedSource ? ` asserted by ${data.trust} content` : ""}: ${names}` +
        (sensitive ? " paired with a privileged request" : ""),
      severity: untrustedSource ? "critical" : "high",
      action: untrustedSource ? "block" : this.actionOnHit,
      spans,
      evidence: { claims, sensitiveRequest: sensitive, authenticated, trust: data.trust },
    });
  }
}

// ---------------------------------------------------------------------------
// Consistency / trajectory gate
// ---------------------------------------------------------------------------

const REFUSAL =
  /\b(?:i\s*(?:'m|\s+am)\s+(?:sorry|afraid|unable)|i\s+can(?:'t|not)\s+(?:help|assist|provide|comply)|i\s+(?:won't|will\s+not)\s+(?:provide|help)|that\s+(?:would\s+be|is)\s+against\s+(?:my|our)\s+(?:guidelines|policy)|i\s+must\s+decline)\b/i;

const PERSISTENCE =
  /\b(?:but\s+(?:you|i)\s+(?:just|already)|come\s+on|please\s+just|try\s+again|what\s+if\s+i\s+(?:told|said)|let\s*'?s\s+try\s+(?:again|another\s+way)|rephrase|you\s+(?:said|answered)\s+(?:that|it)\s+(?:before|earlier)|continue\s+from\s+where)\b/i;

/** Coarse topic vocabularies for measuring drift. */
const TOPIC_TERMS: Record<string, Set<string>> = {
  weapons: new Set("explosive explosives detonator bomb firearm ammunition grenade ordnance warhead propellant".split(" ")),
  chem_bio: new Set("synthesis synthesize precursor reagent toxin pathogen culture aerosolize nerve agent enrichment".split(" ")),
  cyber_offense: new Set("exploit payload ransomware keylogger rootkit backdoor privilege escalation exfiltrate c2 botnet".split(" ")),
  fraud: new Set("counterfeit laundering untraceable forged phishing skimmer carding mule".split(" ")),
  self_harm: new Set("overdose lethal dose asphyxiation suicide".split(" ")),
};

class TrajectoryDetector extends Detector {
  readonly name = "trajectory";
  readonly stages: Stage[] = ["input"];
  override readonly category = "multi_turn_attack";

  private readonly threshold: number;
  private readonly window: number;
  private readonly minTurns: number;
  private readonly actionOnHit: Action;

  constructor(config: DetectorConfig) {
    super(config);
    this.threshold = config.threshold || 0.65;
    this.window = config.options.window ?? 12;
    this.minTurns = config.options.minTurns ?? 3;
    this.actionOnHit = (config.options.action as Action) ?? "block";
  }

  detect(data: DetectorInput): Finding {
    const history = data.history.slice(-this.window);
    const userTurns = history.filter((m) => m.role === "user");
    const assistantTurns = history.filter((m) => m.role === "assistant");

    if (userTurns.length < this.minTurns) {
      return this.clean("conversation too short for trajectory analysis", { turns: userTurns.length });
    }

    const refusals = assistantTurns.filter((m) => REFUSAL.test(m.content)).length;
    const persistence = userTurns.filter((m) => PERSISTENCE.test(m.content)).length;
    const refusalPressure = this.refusalPressure(refusals, persistence, userTurns.length);

    const { topic, score: drift, series } = this.topicDrift([
      ...userTurns,
      { role: "user", content: data.payload } as Message,
    ]);
    const decomposition = this.decomposition(userTurns, topic);

    const signals = { refusal_pressure: refusalPressure, topic_drift: drift, decomposition };
    let score = Math.max(...Object.values(signals));
    // Two co-occurring moderate signals are the documented multi-turn shape.
    if (Object.values(signals).filter((v) => v >= 0.4).length >= 2) score = Math.min(1, score + 0.2);

    const evidence = {
      signals: Object.fromEntries(Object.entries(signals).map(([k, v]) => [k, Number(v.toFixed(3))])),
      refusals,
      persistenceMarkers: persistence,
      turnsAnalysed: userTurns.length,
      driftTopic: topic,
      topicDensityPerTurn: series,
    };

    if (score >= this.threshold) {
      const dominant = (Object.keys(signals) as (keyof typeof signals)[]).reduce((a, b) =>
        signals[a] >= signals[b] ? a : b,
      );
      return this.hit({
        score,
        summary:
          `multi-turn attack pattern (${dominant}) across ${userTurns.length} turns` +
          (topic ? ` converging on '${topic}'` : ""),
        severity: "high",
        action: this.actionOnHit,
        evidence,
      });
    }
    if (score >= this.threshold * 0.6) {
      return this.hit({
        score,
        summary: "conversation shows early escalation signals",
        severity: "low",
        action: "flag",
        evidence,
      });
    }
    return this.clean("trajectory nominal", evidence);
  }

  /** A single refusal is normal; repeated refusals the user pushes against are not. */
  private refusalPressure(refusals: number, persistence: number, turns: number): number {
    if (!refusals) return 0;
    let score = Math.min(1, (refusals / Math.max(1, turns)) * 1.5) * 0.6;
    if (persistence) score += Math.min(0.4, 0.15 * persistence);
    return Math.min(1, score);
  }

  private topicDrift(turns: Message[]): { topic: string | null; score: number; series: number[] } {
    const densities: Record<string, number[]> = {};
    for (const name of Object.keys(TOPIC_TERMS)) densities[name] = [];

    for (const message of turns) {
      const words = message.content.toLowerCase().match(/[a-z]+/g) ?? [];
      const total = Math.max(1, words.length);
      for (const [name, vocab] of Object.entries(TOPIC_TERMS)) {
        densities[name].push(words.filter((w) => vocab.has(w)).length / total);
      }
    }

    let bestTopic: string | null = null;
    let bestScore = 0;
    let bestSeries: number[] = [];

    for (const [topic, series] of Object.entries(densities)) {
      if (!series.some((v) => v > 0)) continue;
      const half = Math.max(1, Math.floor(series.length / 2));
      const firstHalf = series.slice(0, half);
      const secondHalf = series.slice(half).length ? series.slice(half) : series.slice(-1);
      const start = firstHalf.reduce((a, b) => a + b, 0) / firstHalf.length;
      const end = secondHalf.reduce((a, b) => a + b, 0) / secondHalf.length;
      if (end <= start) continue;

      const rise = Math.min(1, (end - start) * 40);
      const magnitude = Math.min(1, end * 30);
      let score = Math.min(1, 0.5 * rise + 0.5 * magnitude);
      // A benign opening that drifts is the attack shape; starting on the
      // topic is just an on-topic conversation.
      if (start < 0.005) score = Math.min(1, score + 0.15);

      if (score > bestScore) {
        bestTopic = topic;
        bestScore = score;
        bestSeries = series.map((v) => Number(v.toFixed(4)));
      }
    }
    return { topic: bestTopic, score: bestScore, series: bestSeries };
  }

  private decomposition(userTurns: Message[], topic: string | null): number {
    if (!topic || userTurns.length < 3) return 0;
    const questions = userTurns.map((m) => m.content).filter((c) => c.trim().endsWith("?"));
    if (questions.length < 3) return 0;
    const lengths = questions.map((q) => q.split(/\s+/).length);
    const mean = lengths.reduce((a, b) => a + b, 0) / lengths.length;
    if (mean > 25) return 0; // long discursive questions are not decomposition
    const variance = lengths.reduce((acc, n) => acc + (n - mean) ** 2, 0) / lengths.length;
    const uniformity = 1 / (1 + variance / Math.max(1, mean));
    return Math.min(1, 0.4 + 0.5 * uniformity * Math.min(1, questions.length / 5));
  }
}

register("authority_claim", (config) => new AuthorityClaimDetector(config));
register("trajectory", (config) => new TrajectoryDetector(config));
export { AuthorityClaimDetector, TrajectoryDetector };
